### PR TITLE
Adds zipkin-sender-stackdriver

### LINKED ISSUE #59: Add a stackdriver-reporter module (so apps can avoid the proxy)
I believe we should be able to make a BytesEncoder<zipkin2.Span> which translates the zipkin span into an opaque byte array representing com.google.devtools.cloudtrace.v1.Trace

This will allow us with a little massaging to use the existing AsyncReporter with a stackdriver sender component.

### PR DESCRIPTION
**Pull request by adriancole**
This adds zipkin-sender-stackdriver, which can be used like so:

```java
  // reads env GOOGLE_APPLICATION_CREDENTIALS just for testing
  @Bean Credentials googleCredentials() throws IOException {
    return GoogleCredentials.getApplicationDefault()
        .createScoped(Collections.singletonList("https://www.googleapis.com/auth/trace.append"));
  }

  @Bean Sender stackdriverSender(Credentials credentials) {
    return StackdriverSender.newBuilder()
        .projectId("zipkin-demo")
        .callOptions(CallOptions.DEFAULT.withCallCredentials(MoreCallCredentials.from(credentials)))
        .build();
  }

  @Bean BytesEncoder<Span> spanBytesEncoder() {
    return StackdriverEncoder.V1;
  }
```

This can drop into any zipkin-reporter enabled app (Sleuth, Brave, Wingtips...)

Fixes #59

### CODE REVIEW

**Comment by adriancole** [comment]
cc @marcingrzejszczak @saturnism

**Comment by adriancole** [comment]
screen shot 2018-03-20 at 4 23 00 pm

**Comment by adriancole** [comment]
OK I've rejigged this.. making the implementation of collector independent from the sender (except sharing "call" types. This is from a suggestion by @anuraaga to not route through byte arrays unless we must (we currently only need to do this in the async reporter)

**Comment by adriancole** [comment]
last steps on my end are to do the lazy cat thing on the sender side. after that I'll put in benchmarks.

an aside but I think we should consider making a bom as actual use of this is very very classpath sensitive. For example, the bom can claim versions of grpc boringssl grpc-auth. Otherwise, project config is a bit complex. This isn't to obviate the google starter, just in case someone wants to use the sender independently.. even without spring boot.. well the maven config is pretty ugly without a bom.

 cc @shakuzen @marcingrzejszczak

**Comment by adriancole** [comment]
ps here's why I think a bom might be helpful.. to get the sleuth example to work I needed to align all of the below (on top of normal sleuth)

```diff
+    <!-- must match https://github.com/grpc/grpc-java/blob/master/SECURITY.md#netty -->
+    <grpc.version>1.10.0</grpc.version>
+    <netty-tcnative-boringssl-static.version>2.0.7.Final</netty-tcnative-boringssl-static.version>
   </properties>
 
   <dependencies>
@@ -34,10 +38,53 @@
       <groupId>org.springframework.cloud</groupId>
       <artifactId>spring-cloud-sleuth-zipkin</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.zipkin.gcp</groupId>
+      <artifactId>zipkin-sender-stackdriver</artifactId>
+      <version>0.1.2-SNAPSHOT</version>
+    </dependency>
+
+    <!-- All the below is likely better served with a BOM -->
+    <dependency>
+      <groupId>com.google.auth</groupId>
+      <artifactId>google-auth-library-oauth2-http</artifactId>
+      <version>0.9.0</version>
+      <exclusions>
+        <exclusion>
+          <groupId>com.google.guava</groupId>
+          <artifactId>guava-jdk5</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+    <!-- actually plumb a transport -->
+    <dependency>
+      <groupId>io.grpc</groupId>
+      <artifactId>grpc-netty</artifactId>
+      <version>${grpc.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-tcnative-boringssl-static</artifactId>
+      <version>${netty-tcnative-boringssl-static.version}</version>
+    </dependency>
+
+    <!-- for MoreCallCredentials -->
+    <dependency>
+      <groupId>io.grpc</groupId>
+      <artifactId>grpc-auth</artifactId>
+      <version>${grpc.version}</version>
+    </dependency>
```

File: sender-stackdriver/src/main/java/zipkin2/reporter/stackdriver/StackdriverMessageSizer.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.List;
+
+import static com.google.protobuf.CodedOutputStream.computeUInt32SizeNoTag;
+
+/** Mutable to avoid array allocation when computing the size of a message */
+final class StackdriverMessageSizer {
+  ArrayList<byte[]> sortedEncodedTraces = new ArrayList<>();
+  byte[] currentTraceId = new byte[32], lastTraceId = new byte[32];
+
+  int messageSizeInBytes(int projectIdInBytes, List<byte[]> encodedTraces) {
```
[the remark this answers is not in the document]
Comment (anuraaga) [reply sender-stackdriver/src/main/java/zipkin2/reporter/stackdriver/StackdriverMessageSizer.java]: Also maybe name the method `serializedSize` or something, the word bytes seems to not be all that related.

File: sender-stackdriver/src/main/java/zipkin2/reporter/stackdriver/StackdriverMessageSizer.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  // bytes should start with a trace_id field, the length of trace ID is always 32, so we know the offset.
+  static final int TRACE_ID_OFFSET = 1 /* field no */ + 1 /* size to encode the number 32 */;
+  static final Comparator<byte[]> TRACE_ID_COMPARATOR = new Comparator<byte[]>() {
+    @Override public int compare(byte[] left, byte[] right) {
+      for (int i = TRACE_ID_OFFSET, length = TRACE_ID_OFFSET + 32; i < length; i++) {
+        int leftByte = left[i] & 0xff;
+        int rightByte = right[i] & 0xff;
+        if (leftByte == rightByte) continue;
+        return leftByte < rightByte ? -1 : 1;
```
Comment (anuraaga) [inline sender-stackdriver/src/main/java/zipkin2/reporter/stackdriver/StackdriverMessageSizer.java]: Guava has `UnsignedBytes.compare` if you want to use that

**Comment by adriancole** [comment]
Thanks. hopefully next commit got it all

**Comment by adriancole** [comment]
ok I have things efficient now. Last step is to merge logic repeated in StackdriverMessageSizer and StackdriverTraceCollator

**Comment by adriancole** [comment]
Ok everything is tested manually. I will merge and release this, but not announce it.

We are <1.0 anyway, and this will allow @saturnism and @anuraaga to try and use it. In case we need to change something, better to delay announcement until then.

**Comment by saturnism** [comment]
Thank you Adrian! Will take this for a spin in the coming weeks (I'll be OOO next week :)

**Pull request merged**